### PR TITLE
crypto: Add a standalone integration test for the NSE race

### PR DESCRIPTION
**Pull request by andybalaam**
Adds a test that fails before https://github.com/matrix-org/matrix-rust-sdk/pull/3338 and passes afterwards.

Tests the fix for https://github.com/matrix-org/matrix-rust-sdk/issues/3110

The test sets up a cross-process lock and 2 clients using it, then makes sure they both have Olm sessions with a 3rd client, before sending from one of them and checking that the other can still receive. This validates that it flushed and replaced its Olm session with the 3rd client.

### CODE REVIEW

**Comment by andybalaam** [comment]
Undrafted to get integration test

**Comment by Hywan** [comment]
Do you need help?

**Comment by andybalaam** [comment]
> Do you need help?

Thank you, no I think I'm nearly there - Rich and I made a breakthough yesterday.

**Comment by andybalaam** [comment]
Requesting review from @Hywan since he was initially chosen.

This is (finally) ready for a first look from you, thanks!

**Review by Hywan — APPROVED** [review]

File: crates/matrix-sdk-crypto/src/olm/account.rs [hunk 1]
```diff
@@ -1277,7 +1277,7 @@ impl Account {
                         // session in lieu of an OTK.)
 
                         warn!(
-                            session_id = session.session_id(),
+                            ?session,
```
Comment (bnjbvr) [inline crates/matrix-sdk-crypto/src/olm/account.rs:1280]: Do you want to keep this logging change?

**Review by bnjbvr — COMMENTED** [review]
(small driveby, sorry)

File: crates/matrix-sdk-crypto/src/olm/account.rs (hunk 1 above)
Comment (andybalaam) [reply crates/matrix-sdk-crypto/src/olm/account.rs]: Thank you! Corrected and re-pushed along with a compile fix.

**Pull request merged**